arangod/Replication2/Supervision/CollectionGroupSupervision.cpp

```diff
@@ -629,6 +714,75 @@ struct TransactionBuilder {
               .end();
   }
 
+  void operator()(RemoveCollectionIndexPlan const& action) {
+    auto tmp = env.write();
+    // Just overwrite all indexes as defined by the Action
+    tmp = std::move(tmp).erase_object(
+        basics::StringUtils::concatT("/arango/Plan/Collections/", database, "/",
+                                     action.cid, "/indexes"),
+        [&](VPackBuilder& builder) { builder.add(action.index.slice()); });
+    env = std::move(tmp)
+              .inc("/arango/Plan/Version")
+              .precs()
+              .isNotEmpty(basics::StringUtils::concatT(
+                  "/arango/Plan/Collections/", database, "/", action.cid))
+              .end();
+  }
+
+  void operator()(AddCollectionIndexPlan const& action) {
+    auto tmp = env.write();
+    // Just overwrite all indexes as defined by the Action
+    tmp = std::move(tmp).push_object(
+        basics::StringUtils::concatT("/arango/Plan/Collections/", database, "/",
+                                     action.cid, "/indexes"),
+        [&](VPackBuilder& builder) {
+          arangodb::velocypack::Slice indexData{action.index->data()};
+          // We need to add the isBuildingFlag
+          TRI_ASSERT(indexData.isObject());
+          VPackObjectBuilder guard{&builder};
+          builder.add(VPackObjectIterator(indexData));
+          if (action.useIsBuilding) {
+            builder.add(StaticStrings::IndexIsBuilding, VPackValue(true));
+          }
+        });
+    env = std::move(tmp)
+              .inc("/arango/Plan/Version")
+              .precs()
+              .isNotEmpty(basics::StringUtils::concatT(
+                  "/arango/Plan/Collections/", database, "/", action.cid))
```

Comment:
I am a little unsure of this precondition. It certainly helps, but is it enough? Note that below, for `UpdateCollectionShardMap` we also check `/arango/Target/CollectionGroups/`.
Is it possible the have a race in the supervision, where we end up with an incomplete collection entry in `Plan`, which leads to a shutdown hanger (impossibility to drop database).
Example:
1. Supervision is requested to add an index to plan.
2.a. It checks the precondition - collection exists.
2.b. In parallel, a different transaction removes the collection.
3. Adds the index entry. We now end up with a broken collection entry in Plan.

I am not certain if this is the case, as it depends on how the supervision handles the snapshots on which it checks preconditions. Maybe someone with more Agency experience drop a hint?